spark/src/main/scala/org/apache/spark/sql/delta/actions/TableFeatureSupport.scala

```diff
@@ -254,6 +317,32 @@ trait TableFeatureSupport { this: Protocol =>
     }
   }
 
+  /**
+   * Remove writer feature from protocol. To remove a writer feature we only need to
+   * remove it from the writerFeatures set.
+   */
+  private[delta] def removeWriterFeature(targetWriterFeature: WriterFeature): Protocol = {
+    require(targetWriterFeature.isRemovable)
+    copy(writerFeatures = writerFeatures.map(_ - targetWriterFeature.name))
+  }
+
+  /**
+   * Remove feature wrapper for removing either Reader/Writer or Writer features. We assume
+   * the feature exists in the protocol. There is a relevant validation at
+   * [[AlterTableRemoveFeatureDeltaCommand]].
+   */
+  def removeFeature(targetFeature: TableFeature): Protocol = targetFeature match {
+    case f: ReaderWriterFeature =>
+      throw new UnsupportedOperationException(
+        "ReaderWriter feature removal is currently not supported")
```

Comment:
I'm not sure it's very helpful to users to tell them about legacy vs. reader-writer etc. All they care is that the feature either can (or cannot) be dropped. If so, this whole match would simplify down to just one `WriterFeature with RemovableFeature` case, plus the default case that throws some appropriate error class?

(plus optional comments explaining that legacy and/or reader-writer commands are still todo)?

Reply:
Added a single message with an appropriate error class.